### PR TITLE
Add optional parameter to showOptions called newEditorGroup

### LINKED ISSUE #106973: viewColumn assignment won't open in rightmost position
- VSCode Version: 1.49.0
- OS Version: Windows 10

Steps to Reproduce:
1. Open two files in separate editors in VS Code
2. In extension development code, utilize the VS Code API to designate extension behavior to open a new editor (of type webViewPanel) in the specified viewColumn. 
3. Specify the viewColumn = 1+ (current number of editors open)

Expected: The new editor will open in a new, right-most column.

Actual: The new editor opens as a tab inside a new column, in between the two open editors.  

Does this issue occur when all extensions are disabled?: Yes/No

I am currently trying to use the VS Code API for extension development. The behavior I am trying to achieve is to open a new webViewPanel editor on the rightmost column on the screen. I have tried assigning the viewColumn of the new editor to all available options shown in the API, but the editor opens in between the currently opened editors. 

https://code.visualstudio.com/api/references/vscode-api#ViewColumn

### PR DESCRIPTION
**Pull request by tomerstav**
The issue was that when a position for a view column that does not exist was converted into editor group it would be marked as side group. Side groups create a new column based on the neighbour of the active group you are on. If it does not exist it will create a new group. So for example if you had two view columns open and you clicked the first one then the first view column is the active group. The second view column is the neighbour of the active group. And the new column would open inside the second view column. But if you clicked the second view column. The active group is the second view column. The neighbour doesn't exist. So the new view column will be a new third group. 

My implementation adds an optional parameter to showOptions called newEditorGroup. When set to true, a new view column is created in a new editor group at the position specified. 

Examples:

1. Two view columns open in their own editor groups. The command below will create a new view column with a new editor group at position 1 resulting in three editor groups.

`const panel = vscode.window.createWebviewPanel("", "", {viewColumn:1, newEditorGroup: true}); `

2. Two view columns open in their own editor groups. The command below will create a new view column with a new editor group at position 3 resulting in three editor groups.

`const panel = vscode.window.createWebviewPanel("", "", {viewColumn:3, newEditorGroup: true}); `

* If view column does not exist, a new editor group is created after the last view column. Which happens in the second example.

This PR fixes #106973

### CODE REVIEW

**Comment by mjbvz** [comment]
I suspect this is not the correct fix. 

`vscode.ViewColumn.ONE === 1` but I believe the internal `EditorViewColumn` type is zero based (so the first column === 0)

**Review by jrieken — CHANGES_REQUESTED** [review]
This seems wrong, please explain those changes

**Comment by tomerstav** [comment]
> This seems wrong, please explain those changes 

I misunderstood how the code was working. I initially thought that the implementation for adding a new view col was expecting non zero based indexing as explained in the docs. https://code.visualstudio.com/api/references/vscode-api#ViewColumn 

My changes fixed the issue but I can see from the failed integration testing and comments that this isn't the right solution. I will convert this pr into a draft and attempt this again.

**Comment by tomerstav** [comment]
I have updated the description of the PR. Please read and let me know if you have any questions. @jrieken @mjbvz

File: src/vs/workbench/services/editor/browser/editorService.ts [hunk 1]
```diff
@@ -681,6 +686,11 @@ export class EditorService extends Disposable implements EditorServiceImpl {
 		return neighbourGroup;
 	}
 
+	private newSideRightGroup(): IEditorGroup {
+		const lastGroupIndex = this.editorGroupService.groups.length - 1;
+		return this.editorGroupService.addGroup(this.editorGroupService.groups[lastGroupIndex], GroupDirection.RIGHT);
```
Comment (bpasero) [inline src/vs/workbench/services/editor/browser/editorService.ts:691]: The group direction should not be hardcoded but taken from user settings via:

```ts
const direction = preferredSideBySideGroupDirection(this.configurationService);
```

File: src/vs/workbench/services/editor/browser/editorService.ts [hunk 2]
```diff
@@ -681,6 +686,11 @@ export class EditorService extends Disposable implements EditorServiceImpl {
 		return neighbourGroup;
 	}
 
+	private newSideRightGroup(): IEditorGroup {
+		const lastGroupIndex = this.editorGroupService.groups.length - 1;
```
Comment (bpasero) [inline src/vs/workbench/services/editor/browser/editorService.ts:690]: That is not a safe way to find the last group, I think you want:

```ts
this.editorGroupService.getGroups(GroupsOrder.GRID_APPEARANCE)
```

File: src/vs/workbench/services/editor/browser/editorService.ts [hunk 3]
```diff
@@ -681,6 +686,11 @@ export class EditorService extends Disposable implements EditorServiceImpl {
 		return neighbourGroup;
 	}
 
+	private newSideRightGroup(): IEditorGroup {
```
Comment (bpasero) [inline src/vs/workbench/services/editor/browser/editorService.ts:689]:
```suggestion
	private newSideGroup(): IEditorGroup {
```

File: src/vs/workbench/services/editor/browser/editorService.ts [hunk 4]
```diff
@@ -614,6 +614,11 @@ export class EditorService extends Disposable implements EditorServiceImpl {
 			targetGroup = this.findSideBySideGroup();
 		}
 
+		// Group: New group
+		else if (group === NEW_GROUP) {
+			targetGroup = this.newSideRightGroup();
```
Comment (bpasero) [inline src/vs/workbench/services/editor/browser/editorService.ts:619]:
```suggestion
			targetGroup = this.newSideGroup();
```

File: src/vs/workbench/services/editor/browser/editorService.ts [hunk 5]
```diff
@@ -42,7 +42,7 @@ import { IModelService } from 'vs/editor/common/services/modelService';
 import { ILogService } from 'vs/platform/log/common/log';
 
 type CachedEditorInput = ResourceEditorInput | IFileEditorInput | UntitledTextEditorInput;
-type OpenInEditorGroup = IEditorGroup | GroupIdentifier | SIDE_GROUP_TYPE | ACTIVE_GROUP_TYPE;
+type OpenInEditorGroup = IEditorGroup | GroupIdentifier | SIDE_GROUP_TYPE | ACTIVE_GROUP_TYPE | NEW_GROUP_TYPE;
```
Comment (bpasero) [inline src/vs/workbench/services/editor/browser/editorService.ts:45]: I find `NEW_GROUP_TYPE` is not fully expressing what it does. Based on the implementation it seems to create a new group after the last group by appearance. If so I think the name should reflect that, because as a user of the API one might think that it opens a new group from the currently active group and not the last group.

**Review by bpasero — CHANGES_REQUESTED** [review]
Overall I think this can break existing extensions because we used to return `SIDE_GROUP` which is defined as opening to the side of the currently active group and `NEW_GROUP` is entirely different, opening a new group after the last group by appearance.

**Comment by tomerstav** [comment]
> Overall I think this can break existing extensions because we used to return `SIDE_GROUP` which is defined as opening to the side of the currently active group and `NEW_GROUP` is entirely different, opening a new group after the last group by appearance.

I think another approach that wouldn't break existing extensions would be to add a new optional boolean property called newEditorGroup to WebviewPanelOptions. This feature would allow you to create webView columns in new editor groups if the property newEditorGroup was passed in. 

Depending on the directions in settings and the viewCol number passed in, you could control where this editor group would be created. In the issue #106973, if I had two view columns and I wanted to open a new third view column in its own editor group at the end, I would use vscode.window.createWebviewPanel(“”,””,2, {newEditorGroup: true}); Then regardless of my active editor I could create a third view col at the end in its own new editor group. I am assuming I have direction as Right, which I believe is default.

Please let me know what you think of this idea.

File: src/vs/workbench/services/editor/browser/editorService.ts (hunk 3 above)
Comment (tomerstav) [reply src/vs/workbench/services/editor/browser/editorService.ts:689]: See my comment below explaining a new approach. But could change it to simply newEditorGroup.

File: src/vs/workbench/services/editor/browser/editorService.ts (hunk 5 above)
Comment (tomerstav) [reply src/vs/workbench/services/editor/browser/editorService.ts:45]: I will remove this with new approach.

File: src/vs/workbench/services/editor/browser/editorService.ts (hunk 4 above)
Comment (tomerstav) [reply src/vs/workbench/services/editor/browser/editorService.ts:619]: See comment above

**Comment by tomerstav** [comment]
> > Overall I think this can break existing extensions because we used to return `SIDE_GROUP` which is defined as opening to the side of the currently active group and `NEW_GROUP` is entirely different, opening a new group after the last group by appearance.
> 
> I think another approach that wouldn't break existing extensions would be to add a new optional boolean property called newEditorGroup to WebviewPanelOptions. This feature would allow you to create webView columns in new editor groups if the property newEditorGroup was passed in.
> 
> Depending on the directions in settings and the viewCol number passed in, you could control where this editor group would be created. In the issue #106973, if I had two view columns and I wanted to open a new third view column in its own editor group at the end, I would use vscode.window.createWebviewPanel(“”,””,2, {newEditorGroup: true}); Then regardless of my active editor I could create a third view col at the end in its own new editor group. I am assuming I have direction as Right, which I believe is default.
> 
> Please let me know what you think of this idea. @bpasero
> 
> TLDR: viewCol acts as pointer and newEditorGroup states whether to create a new editor group for the view col

@mjbvz @jrieken could either of you please comment on whether this is the correct approach?

**Review by bpasero — CHANGES_REQUESTED** [review]
Why is `newEditorGroup` a boolean and not another type similar to `SIDE_GROUP_TYPE`? I would rather introduce something like `NEW_GROUP_TYPE`.

**Comment by tomerstav** [comment]
> Why is `newEditorGroup` a boolean and not another type similar to `SIDE_GROUP_TYPE`? I would rather introduce something like `NEW_GROUP_TYPE`.

We need to keep track of the view column number. Groups internally was keeping track of that number. I have create a NEW_GROUP_TYPE and converted newEditorGroup to keep track of the view column number instead.

**Comment by bpasero** [comment]
It looks like https://github.com/microsoft/vscode/issues/106973 is now closed with a workaround and I am not happy with the change in this PR to be honest. I would suggest to not push this forward.

**Review by bpasero — CHANGES_REQUESTED** [review]
Rejected as per https://github.com/microsoft/vscode/pull/107555#issuecomment-707552637

**Comment by tomerstav** [comment]
> It looks like #106973 is now closed with a workaround and I am not happy with the change in this PR to be honest. I would suggest to not push this forward.

Could you explain why you are not happy with this PR? I would like to learn from this.

**Comment by bpasero** [comment]
I actually started to review this PR but then ended up changing a lot of the code myself, almost rewriting it. In principle I am not opposed to the idea of having a new group type in editor land, but it touched too many areas in the end and I am not convinced we actually still need this?

**Comment by tomerstav** [comment]
> I actually started to review this PR but then ended up changing a lot of the code myself, almost rewriting it. In principle I am not opposed to the idea of having a new group type in editor land, but it touched too many areas in the end and I am not convinced we actually still need this?

I think the work around is not a good long term solution. But fair enough I will close the PR.
